File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.java
import java.io.*;
import java.util.*;

public class Main {
    
    static int n;
    static int[][] board = new int[22][22];
    static int[][] board2 = new int[22][22];
    static int max = 0;

    public static void updateMaxBlock() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                max = Math.max(max, board2[i][j]);
            }
        }
    }

    public static void rotateSticker() {
        int[][] result = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                result[j][n - i - 1] = board2[i][j];
            }
        }
        board2 = result;
    }

    public static void move(int dir) {
        while (dir-- > 0) rotateSticker();
        boolean[][] merged = new boolean[n + 1][n + 1];
        for (int i = 0; i < n; i++) {
            for (int j = 1; j < n; j++) {
                if (board2[i][j] == 0) continue;
                int idx = j - 1;        // 한 칸 왼쪽부터 검색
                boolean moved = false;
        
                // 1) 왼쪽으로 빈 칸이 있는지, 아니면 병합 대상이 있는지 탐색
                while (idx >= 0 && board2[i][idx] == 0) {
                    idx--;
                }
        
                // 2) idx가 -1이면 끝까지 빈 칸만 있었으므로, “맨 왼쪽(0칸)”으로 이동
                if (idx < 0) {
                    board2[i][0] = board2[i][j];
                    board2[i][j] = 0;
                    moved = (j != 0);
                }
                else {
                    // 3) idx가 0..n-1 까지 빈 칸이 아닌 블록을 가리키는 상태
                    if (board2[i][idx] == board2[i][j] && !merged[i][idx]) {
                        // 병합
                        board2[i][idx] *= 2;
                        merged[i][idx] = true;
                        board2[i][j] = 0;
                        moved = true;
                    }
                    else {
                        // 빈 칸이 아닌 블록이랑 값이 다르면, idx+1 위치로 한 칸만 이동
                        if (idx + 1 != j) {
                            board2[i][idx + 1] = board2[i][j];
                            board2[i][j] = 0;
                            moved = true;
                        }
                    }
                }
                // (moved 여부에 따라 0으로 덮거나, 덮지 않아도 되지만, 위에서 이미 처리함)
            }
        }
    }

    public static void run() {
        for (int tmp = 0; tmp < (1 << 10); tmp++) {
            for (int i = 0; i < n; i++) {
                board2[i] = board[i].clone();
            }
            int brute = tmp;
            for (int k = 0; k < 5; k++) {
                int dir = brute % 4;
                brute /= 4;
                move(dir);
            }
            updateMaxBlock();
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        run();
        System.out.print(max);

    }

    static class Pair {
        int x; int y;
        public Pair(int xx, int yy) {
            x = xx;
            y = yy;
        }
    }
}

File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.kt
/**
 * 규칙
 * - 같은 값을 갖는 두 블록이 충돌하면 두 블록은 하나로 합쳐지게 된다
 * - 한 번의 이동에서 이미 합쳐진 블록은 또 합쳐질 수 없다.
 * - 똑같은 수가 세 개가 있는 경우에는 이동하려고 하는 쪽의 칸이 먼저 합쳐진다.
 * 구하고자 하는 것
 * - 최대 5번 이동해서 만들 수 있는 가장 큰 블록의 값
 */
import java.util.*

val dx = intArrayOf(1, 0, -1, 0)
val dy = intArrayOf(0, 1, 0, -1)
fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    var board = Array(n) { StringTokenizer(readLine()).let { st -> IntArray(n) { st.nextToken().toInt() } } }

    fun moveHorizon(dir: Int) {
        val collision = Array(n) { BooleanArray(n) }
        val (sy, ey) = if (dir == 1) (n - 1 to 0) else (0 to n - 1)
        for (cx in 0 until n) {
            val blockIndexes = mutableListOf<Int>()
            for (cy in IntProgression.fromClosedRange(sy, ey, -dy[dir])) {
                if (board[cx][cy] != 0) blockIndexes.add(cy)
            }

            for (cy in blockIndexes) {
                var ty = cy + dy[dir] // ty는 블록이 합쳐지거나 위치할 곳의 다음 위치
                while ((ty in 0 until n) && (board[cx][ty] == 0)) ty += dy[dir]

                // 벽으로 부딪히는 경우
                if (ty !in 0 until n) {
                    ty = ty - dy[dir]
                    if (ty == cy) continue
                    board[cx][ty] = board[cx][cy]
                    board[cx][cy] = 0
                    continue
                }

                // 기존 블록과 부딪히는 경우
                // 1. 값이 같고, 충돌한 적이 없는 경우
                if (board[cx][ty] == board[cx][cy] && !collision[cx][ty]) {
                    board[cx][ty] *= 2
                    collision[cx][ty] = true
                    board[cx][cy] = 0
                    continue
                }

                // 그 외의 경우
                ty = ty - dy[dir]
                if (ty != cy) {
                    board[cx][ty] = board[cx][cy]
                    board[cx][cy] = 0
                }
            }
        }
    }

    fun moveVertical(dir: Int) {
        val collision = Array(n) { BooleanArray(n) }
        val (sx, ex) = if (dir == 0) (n - 1 to 0) else (0 to n - 1)
        for (cy in 0 until n) {
            val blockIndexes = mutableListOf<Int>()
            for (cx in IntProgression.fromClosedRange(sx, ex, -dx[dir])) {
                if (board[cx][cy] != 0) blockIndexes.add(cx)
            }

            for (cx in blockIndexes) {
                var tx = cx + dx[dir]
                while ((tx in 0 until n) && (board[tx][cy] == 0)) tx += dx[dir]

                // 벽으로 부딪히는 경우
                if (tx !in 0 until n) {
                    tx = tx - dx[dir]
                    if (tx == cx) continue
                    board[tx][cy] = board[cx][cy]
                    board[cx][cy] = 0
                    continue
                }

                // 기존 블록과 부딪히는 경우
                // 1. 값이 같고, 충돌한 적이 없는 경우
                if (board[tx][cy] == board[cx][cy] && !collision[tx][cy]) {
                    board[tx][cy] *= 2
                    collision[tx][cy] = true
                    board[cx][cy] = 0
                    continue
                }

                tx = tx - dx[dir]
                if (tx != cx) {
                    board[tx][cy] = board[cx][cy]
                    board[cx][cy] = 0
                }
            }
        }
    }

    fun moveDir(dir: Int) {
        if (dir == 1 || dir == 3) moveHorizon(dir)
        else moveVertical(dir)
    }

    var max = 0
    fun move(moveCnt: Int) {
        if (moveCnt == 5) {
            val localMax = board.maxOf { line -> line.max() }
            max = maxOf(max, localMax)
            return
        }
        
        for (i in 0 until 4) {
            val originBoard = Array(n) { idx -> board[idx].clone() }
            moveDir(i)
            move(moveCnt + 1)
            board = originBoard
        }
    }

    move(0)
    print(max)
}